File: jiuzhang/substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.py
"""
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
"""

class Solution(object):
    """ 
    O(nm)
    n, m, k = len(s), len(words[0]), len(words)
    暴力做法，枚举开始位置，判断之后长度m*k的子串是否由给定字符串集合组成，最坏复杂度为O(nmk)。
    对于长度为m的字符串，0与m位置开始的区别，只在于少了s[0:m]，多了s[m*k+1:(m+1)*k]，所以产生了许多
    冗余操作。我们根据开始位置0~m-1分类，扫描字符串s，使用一个滑动窗口记录当前匹配了那些字符串，当下一个
    字符串不在words中，清空窗口(任意包含该串的均不合法)，如果记录的出现次数超过了words中数量，表示需要滑动窗口，
    窗口中单词数量等于k时，更新答案。
    """
    def findSubstring(self, s, words):
        hash = {}
        res = []
        wsize = len(words[0])
        
        for str in words:
            if str in hash:
                hash[str] += 1
            else:
                hash[str] = 1
        for start in range(0, len(words[0])):
            slidingWindow = {}
            wCount = 0
            for i in range(start, len(s), wsize):
                word = s[i : i + wsize]
                if word in hash:
                    if word in slidingWindow:
                        slidingWindow[word] += 1
                    else:
                        slidingWindow[word] = 1
                    wCount += 1
                    while hash[word] < slidingWindow[word]:
                        pos = i - wsize * (wCount - 1)
                        removeWord = s[pos : pos + wsize]
                        print i, removeWord
                        slidingWindow[removeWord] -= 1
                        wCount -= 1
                else:
                    slidingWindow.clear()
                    wCount = 0
                if wCount == len(words):
                    res.append(i - wsize * (wCount - 1))
                    
        return res

File: jiuzhang/substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.cpp
/**
 * 本代码由九章算法编辑提供。没有版权欢迎转发。
 * - 九章算法致力于帮助更多中国人找到好的工作，教师团队均来自硅谷和国内的一线大公司在职工程师。
 * - 现有的面试培训课程包括：九章算法班，系统设计班，九章强化班，Java入门与基础算法班
 * - 更多详情请见官方网站：http://www.jiuzhang.com/
 */

// This solution cost only O(n * m) time where n is length of the master string s. m is the length of each word.
// This solution"s algorithm is using sliding window to optimize the traverse of the string.
// directly for every start index and then determine whether this start index is a legal result will cost O(n * m * k) time where k is the numbers of words.

class Solution {
public:
    vector findSubstring(string s, vector<string>& words) {
        vector result;
        if (words.size() == 0) {
            return result;
        }
        
        unordered_map hash;
        for (int i = 0; i < words.size(); i++) {
            if (hash.find(words[i]) == hash.end()) {
                hash[words[i]] = 1;
            } else {
                hash[words[i]]++;
            }
        }
        
        int wSize = words[0].length();
        for (int start = 0; start < wSize; start++) {
            int wCount = 0;
            unordered_map slidingWindow;
            for (int i = start; i + wSize <= s.length(); i+=wSize) {
                string word = s.substr(i, wSize);
                if (hash.find(word) == hash.end()) {
                    slidingWindow.clear();
                    wCount = 0;
                } else {
                    wCount++;
                    if (slidingWindow.find(word) == slidingWindow.end()) {
                        slidingWindow[word] = 1;
                    } else {
                        slidingWindow[word]++;
                    }
                    while (hash[word] < slidingWindow[word]) {
                        string removedWord = s.substr(i - (wCount-1) * wSize, wSize);
                        slidingWindow[removedWord]--;
                        wCount--;
                    }
                }
                
                if (wCount == words.size()) {
                    result.push_back(i - (wCount - 1) * wSize);
                }
            }
        }
        
        return result;
    }
};
